[Aslam] iP

aslam341 opened the pull request:
### David
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

David frees your mind of having to remember things you need to do. It's

- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is,

1. download it from [here](https://github.com/aslam341/ip/releases/tag/A-Jar).
2. move the Jar file to a desired folder.
3. open the command line in that folder and run `java -jar Duke.jar`
4. add your tasks.
5. let it manage your tasks for you :v:

And it is __FREE__!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    String filePath = "./data/duke.txt";
    Duke duke = new Duke(filePath);
    duke.run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aslam341 committed:
Rename chatbot to David

aslam341 committed:
Implement initial skeleton version

aslam341 committed:
Echoes user commands

aslam341 committed:
Use existing function to greet

aslam341 committed:
Add ability to store and display tasks

aslam341 committed:
Add Task class

aslam341 committed:
Added mark and unmark task function

aslam341 committed:
Implement three types of tasks

aslam341 committed:
Add automated text UI test

aslam341 committed:
Handle incorrect inputs by users

aslam341 committed:
Add ability to delete tasks

aslam341 committed:
Update automated text UI test

aslam341 committed:
Add file string format for tasks

aslam341 committed:
Add save tasks to file feature

aslam341 committed:
Add load tasks from file feature

aslam341 committed:
Handle case where data folder or file not found

aslam341 committed:
Merge branch 'branch-Level-7'

aslam341 committed:
Add date and time to deadline and event tasks

aslam341 committed:
Merge branch 'branch-Level-8'

aslam341 committed:
Add Storage and Ui classes

aslam341 committed:
Add Parser and Command classes

aslam341 committed:
Update Storage and Ui class

aslam341 committed:
Update main and tasks to use new classes

aslam341 committed:
Merge branch 'branch-A-MoreOOP'

aslam341 committed:
Shift java files into packages

aslam341 committed:
Merge branch 'branch-A-Packages'

aslam341 committed:
Merge branch 'add-gradle-support' into branch-A-gradle

aslam341 committed:
Change mainClass property

aslam341 committed:
Add skeleton code for unit test

aslam341 committed:
Merge branch 'branch-A-gradle'

aslam341 committed:
Add ParserTest JUnit test

aslam341 committed:
Modify import statement in ParserTest

aslam341 committed:
Add TaskListTest JUnit test

aslam341 committed:
Merge branch 'branch-A-JUnit'

aslam341 committed:
Add JavaDoc to Duke class

aslam341 committed:
Move common DateTimeFormatter-s to Ui class

aslam341 committed:
Merge branch 'master' into branch-A-JavaDoc

aslam341 committed:
Add JavaDoc to Task class

aslam341 committed:
Add JavaDoc to TaskList class

aslam341 committed:
Add JavaDoc to the three child classes of Task

aslam341 committed:
Add JavaDoc to Ui class

aslam341 committed:
Fix JavaDoc of class members

aslam341 committed:
Add JavaDoc to Storage class

aslam341 committed:
Add JavaDoc to Parser and Command classes

aslam341 committed:
Add JavaDoc to DukeException class

aslam341 committed:
Make class names for commands be nouns

aslam341 committed:
Change common prefix for associated date formats

aslam341 committed:
Place line breaks to improve readability

aslam341 committed:
Add ability to find tasks by keyword

aslam341 committed:
Merge branch 'branch-A-JavaDoc'

aslam341 committed:
Resolve merge conflict

aslam341 committed:
Resolve merge conflict

itssisi commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    public boolean isDone() {
+        return this.isDone;
+    }
```
Good naming for boolean as it is phrased in question form.

itssisi commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,148 @@
+     * @return An ArrayList of Task objects loaded from the storage file.
+     * @throws IOException If an error occurs while reading from the file or if the file contains invalid data.
+     */
+    private ArrayList<Task> loadTasksFromFile() throws IOException {
```
Is this method name too complicated?

itssisi commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,148 @@
+     * @param filePath The file path to check or create.
+     * @throws DukeException If an error occurs while creating directories or files.
+     */
+    private void accessOrCreateFile(String filePath) throws DukeException {
```
Is this method name too long?

itssisi commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,127 @@
+     *
+     * @param task The task to be added.
+     */
+    public void addTask(Task task) {
```
Good naming of method using a verb.

itssisi commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,137 @@
+    public static Command parse(String userInput) {
+        String[] userInputArray = userInput.split(" ", 2);
+        String command = userInputArray[0];
+        String commandDetails = userInputArray.length > 1 ? userInputArray[1] : "";
```
Good practice of calculating intermediate values first to avoid lengthy code.

itssisi commented:
I think overall the code has good SE practices in naming convention, but some names sound quite complicated. It would be good if simpler names are used.

aslam341 committed:
Update automated text ui testing

aslam341 committed:
Update EXPECTED.TXT for automated text ui testing

aslam341 committed:
Add CheckStyle and fix violations accordingly

aslam341 committed:
Merge branch 'branch-A-CheckStyle'

aslam341 committed:
Modify build.gradle to set up JavaFX

aslam341 committed:
Change from text UI to GUI

aslam341 committed:
Move GUI files to package

aslam341 committed:
Merge branch 'branch-Level-10'

conradsoon commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,75 @@
+            storage = new Storage(filePath);
+            taskList = new TaskList(storage.load());
+        } catch (DukeException e) {
+            mainWindow.printMessage("Error loading duke.tasks from file: " + e.getMessage());
```
Might be better to do "Error loading tasks from file:"

conradsoon commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,136 @@
+     * @param userInput The user's input command as a string.
+     * @return A command object that corresponds to the parsed input.
+     */
+    public static Command parse(String userInput) {
```
Might it be beneficial to break down the parsing of different commands into their own methods for readability and maintainability?

conradsoon commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,128 @@
+        if (foundTasks.size() == 0) {
+            return "No matching tasks found.";
+        } else {
+            String s = "Here are the matching tasks in your list:";
```
Maybe a more descriptive name, such as taskListString or outputString, could be more appropriate?

conradsoon reviewed (commented):
Code is clear, concise and understandable, and I have learnt a lot looking through it.

aslam341 committed:
Modify build.gradle main class property

aslam341 committed:
Add assert statements to some classes

There are several assumptions that are made when executing certain
steps in the code.

Assertions are used to define assumptions about the program state so
that the runtime can verify them.

As a step towards verifying those assumptions, lets add some assert
statements. Doing so will make it easier to find possible bugs in
the code.

aslam341 committed:
Improve code quality through refactoring

Code quality is good but can be improved further.

Among various dimensions of code quality, such as run-time efficiency,
security, and robustness, one of the most important is readability.

Let's improve the readability of the code by:
- Making the happy path more obvious
- Using the SLAP principle
- Refactoring using the Extract Method technique

aslam341 committed:
Merge pull request #1 from aslam341/branch-A-Assertions

Add assert statements to some classes

aslam341 committed:
Resolve merge conflict

aslam341 committed:
Merge pull request #2 from aslam341/branch-A-CodeQuality

Improve code quality through refactoring

aslam341 committed:
Add priority attribute to tasks

aslam341 committed:
Allow users to modify task priority

aslam341 committed:
Merge branch 'branch-C-Priority'

aslam341 committed:
Fix JavaDoc in a few classes

aslam341 committed:
Fix truncation due to long messages

aslam341 committed:
Fix goodbye message not showing before exiting

aslam341 committed:
Refactor parser class using extract method

aslam341 committed:
Check event start time is before end time

aslam341 committed:
Add more JUnit tests and rename some tests

aslam341 committed:
Set title for the application

aslam341 committed:
Add representative screenshot of product

aslam341 committed:
Enhance 'bye' command handling

aslam341 committed:
Fix line breaks from 'find' command output

aslam341 committed:
Add User Guide

aslam341 committed:
Fix README.md line breaks

aslam341 committed:
Fix README.md quick start section

aslam341 committed:
Trigger rebuild
